Use mount.Target to specify subdirectory of rootfs mount

hinshun opened the pull request:
Fixes #7839

---

- Add Target to mount.Mount
- Add UnmountAllRecursive to unmount deepest mount first for a list of mount.Mount.
- Add UnmountRecursive to unmount deepest mount first for a given target, using /proc/mountinfo.

Signed-off-by: Edgar Lee <edgarhinshunlee@gmail.com>

dmcgowan commented on mount/mount.go:
```diff
@@ -38,3 +45,39 @@ func All(mounts []Mount, target string) error {
+	}
+
+	// Make the deepest mount be first
+	sort.Slice(targets, func(i, j int) bool {
```
Probably best to use `SliceStable` here

dmcgowan commented on mount/mount.go:
```diff
@@ -38,3 +45,39 @@ func All(mounts []Mount, target string) error {
+
+	// Make the deepest mount be first
+	sort.Slice(targets, func(i, j int) bool {
+		return len(targets[i]) > len(targets[j])
```
Not sure you could safely assume this is the unmount order, perhaps using a prefix check rather than len check.

hinshun replied:
There are some places which uses len check as well:
https://github.com/containerd/containerd/blob/v1.7.0-beta.1/mount/temp_unix.go#L48-L51
https://github.com/containerd/containerd/blob/v1.7.0-beta.1/pkg/cri/server/helpers_linux.go#L170-L173

Giving this a second thought... if there is an unmount order that is invalid, that would mean a parent mount was unmounted prior to a submount. By virtue of being a parent mount, its mountpoint is a strict prefix of the mountpoint of the submount, so a len check will be sufficient in this case. If it isn't a strict prefix, then the unmount order shouldn't matter.

dmcgowan commented on mount/mount_freebsd.go:
```diff
@@ -36,7 +37,7 @@ func (m *Mount) Mount(target string) error {
 	// function for FreeBSD, so instead we execute mount(8) and trust it to do
 	// the right thing
-	return m.mountWithHelper(target)
+	return m.mountWithHelper(filepath.Join(target, m.Target))
```
Would it always be an error then to have a non-empty target in a mount list of 1? If this is the first mount, the directory won't exist.

hinshun replied:
Yes it would always error.

Currently in my snapshotter, I prepare a lowerdir in an overlay mount with all the mountpoints already created before chaining a bind mount to a non-empty target. Do you think it should be the runtime's responsibility to ensure the existence of these mountpoints?

dmcgowan reviewed (commented):
I would still like to understand the snapshotter use case better for doing this. I think there are other use cases around mount that should be considered alongside this, especially the chaining mounts (previous mount target has source for later mount) and custom mounts (mounts done through plugin or something other than a mount syscall).

I think this is probably standalone enough to consider on its own, assuming we can make the `Target` non-ambiguous in the interfaces.

hinshun commented:
> I would still like to understand the snapshotter use case better for doing this.

I am writing a remote snapshotter for [nix](https://nixos.org/) to prepare rootfs of nix-based images directly from the local nix store rather than downloading blobs from the registry. The nix store is a readonly filesystem with no collision between installed packages. For example, all the paths necessary to execute `bin/hello` which outputs `Hello World!` are the following paths:

```
/nix/store/34xlpp3j3vy7ksn09zh44f1c04w77khf-libunistring-1.0
/nix/store/5mh5019jigj0k14rdnjam1xwk5avn1id-libidn2-2.3.2
/nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163
/nix/store/g2m8kfw7kpgpph05v2fxcx4d5an09hl3-hello-2.12.1
```

We already upload all this data to S3 in nix's archive format, but then we'll have to repackage them into layers tarballs and upload to a registry to use them in containers. Putting multiple paths in a layer greatly duplicates the data, and separating paths into individual layers run into the 125 layer limit. In the current nix ecosystem, there are layering algorithms that bucket paths into layers based on "popularity", but it is only a heuristic.

In the new nix snapshotter, we have a base overlay snapshotter that will handle compatibility with regular layers and provide a read write filesystem. A new mediatype is introduced that should be remotely unpacked, we then chain a series of bind mounts using remote snapshot annotations from the layer:

```
/nix/store/34xlpp3j3vy7ksn09zh44f1c04w77khf-libunistring-1.0 mount to /<rootfs>/nix/store/34xlpp3j3vy7ksn09zh44f1c04w77khf-libunistring-1.0
...
```

Nix store paths are not relocatable so they need to be mounted in the same path in the container. Since the closure is completely self-contained, that's all that is necessary for it to function within the container.

Since overlayfs doesn't propagate submounts from lowerdirs, this seems to be only option.

hinshun commented:
@dmcgowan I've changed `UnmountAll` to simply unmount in reverse order of the list of mounts provided, and added more documentation in godocs and in `docs/snapshotters/README.md`.

estesp commented:
Overall this looks reasonable; one question: I see no validation of `m.Target` all the places it is now used; do we need to check any bounds/safety conditions (symlinks, relative dir paths, rooted paths?)

dmcgowan commented:
@estesp root joining the target seems like the right thing to do. If the target itself is under a directory which needs to be traversed inside the container, then it should not be traversed without root join.

dmcgowan commented:
Try [`fs.RootPath`](https://pkg.go.dev/github.com/containerd/continuity/fs#RootPath)

thaJeztah commented on mount/mount_unix.go:
```diff
@@ -18,24 +18,41 @@
-	return ErrNotImplementOnUnix
+// UnmountRecursive unmounts the target and all mounts underneath, starting
+// with the deepest mount first.
+func UnmountRecursive(target string, flags int) error {
```
Should we use https://github.com/moby/sys/blob/c161267cd2212985f81a50a029001f82c91bca7f/mount/mount_unix.go#L41-L44 for this? (and/or see if things are missing in that implementation)

hinshun replied:
Looks like we'll need to refactor `UnmountAll` to `moby/sys/mount` if we want to use that. It doesn't do the repeat unmount till EINVAL.

thaJeztah replied:
Do we know the reason for the repeat unmount? (I've seen some code where I was wondering: "was this just a big hammer to fix a problem that wasn't fully understood")?

hinshun replied:
Looks like it came from this discussion from PR:
https://github.com/containerd/containerd/pull/1141#discussion_r127015970

thaJeztah replied:
Thanks for digging that up; yes, that _"feels"_ like a "let's just use a bigger hammer" discussion, and could be worth looking into.

Main motivation for my comment was that I've seen various "subtle" issues with mount/unmount (and more notably; parsing mountinfo). Code like this can easily start spreading around and diverge, and I _think_ it's a worthy investment to have a solid, well-tested, canonical implementation for such things (if possible).

Let me also ping @kolyshkin (who may have some thoughts as well, as he did a lot of digging into subtleties).

I agree with @dmcgowan that that should not be a blocker for this PR though; I think it's fine if we acknowledge this may need to be verified / looked into, and do that separately.

hinshun commented:
Is there anything else blocking this PR? I made changes to use `fs.RootPath` where appropriate.

dmcgowan commented:
I think its good to go now. If we want to address using a another package or changing the unmount behavior, that should be done separately to this PR.

thaJeztah commented on mount/mount_unsupported.go:
```diff
@@ -0,0 +1,46 @@
+}
+
+// UnmountRecursive is not implemented on this platform
+func UnmountRecursive(mount string, flags int) error {
```
Looks like `UnmountRecursive` has a duplicate declaration for `openbsd` (defined both in `_unix.go` and `_unsupported.go`)

hinshun replied:
Thank you, I fixed this and squashed my commits into one.

thaJeztah commented on mount/mount.go:
```diff
@@ -16,6 +16,12 @@
+import (
+	"fmt"
+
+	"github.com/containerd/continuity/fs"
```
I'll probably blame myself if I don't comment; was looking if this would be an issue to add as new dependency for this package (and if another approach would work). Looks like continuity was already a dependency (but for testing only). I didn't see a good alternative immediately though (resolving the path before calling the `mount` package didn't seem like an immediate option, so "probably" fine.

thaJeztah reviewed (approved):
LGTM, thanks!

hinshun commented:
Thanks for the review. I double checked my work against the [CONTRIBUTING](https://github.com/containerd/project/blob/main/CONTRIBUTING.md) guide and validated commit 34d5878 with my snapshotter. Anything else to do?
